### PR TITLE
EN 3274 reactivated tx nonce check on transactions

### PR DESCRIPTION
**Pull request by iulianpascalau**
reactivated tx nonce check on transactions
modified txInterceptor and header interceptors not to check the storer for a fetched item
added interfaces and nil (mock) implementations for newly received items
unskipped tx nonce check unit tests
fixed integration tests that used tx processing

### CODE REVIEW

File: integrationTests/multiShard/block/executingMiniblocksSc_test.go [hunk 1]
```diff
@@ -308,7 +313,9 @@ func TestShouldProcessWithScTxsJoinNoCommitShouldProcessedByValidators(t *testin
 	syncBlock(t, nodes, idxProposers, round)
 	round = incrementAndPrintRound(round)
 
-	nodeJoinsGame(nodes, idxProposerShard0, topUpValue, hardCodedScResultingAddress)
+	nonceNodeIdx0 := uint64(0)
```
Comment (sasurobert) [inline integrationTests/multiShard/block/executingMiniblocksSc_test.go]: this has to be refactored together with the whole file

File: integrationTests/multiShard/block/executingMiniblocksSc_test.go [hunk 2]
```diff
@@ -688,13 +699,14 @@ func createTxDeploy(
...
 	topUpVal *big.Int,
 	scAddress []byte,
 ) *transaction.Transaction {
 
 	tx := &transaction.Transaction{
-		Nonce:    0,
+		Nonce:    nonce,
 		Value:    topUpVal,
 		RcvAddr:  scAddress,
 		SndAddr:  tn.OwnAccount.PkTxSignBytes,
```
Comment (sasurobert) [inline integrationTests/multiShard/block/executingMiniblocksSc_test.go]: OwnAccount has Balance as field and it can get updated separately to see if the value in the blockchain is the same. This is done for the singleShard SC test and for the long tests. The same could be done with the Nonce. Add the Nonce field to the TestWallettAccount structure, and increase its value there. Then you can eliminate the nonce uint64 as a parameter from a lot of functions.

File: integrationTests/multiShard/block/executingMiniblocks_test.go [hunk 3]
```diff
@@ -373,18 +382,45 @@ func TestShouldProcessBlocksInMultiShardArchitecture(t *testing.T) {
 
 }
 
+func getBlockBody(
```
Comment (sasurobert) [inline integrationTests/multiShard/block/executingMiniblocks_test.go]: function does not explain what it does.

File: integrationTests/multiShard/block/executingMiniblocks_test.go [hunk 4]
```diff
@@ -118,7 +127,7 @@ func TestShouldProcessBlocksInMultiShardArchitecture(t *testing.T) {
 			err := n.blkProcessor.ProcessBlock(
 				n.blkc,
 				n.headers[0],
-				block.Body(n.miniblocks),
+				getBlockBody(testMarshalizer, testHasher, n.headers[0], n.miniblocks),
```
Comment (sasurobert) [inline integrationTests/multiShard/block/executingMiniblocks_test.go]: why is this needed ? n.miniblocks should only have the miniblocks to this block.

File: process/block/interceptors/headerInterceptor.go [hunk 5]
```diff
@@ -80,10 +80,9 @@ func (hi *HeaderInterceptor) processHeader(hdrIntercepted *block.InterceptedHead
 		return
 	}
 
-	err := hi.hdrInterceptorBase.storer.Has(hdrIntercepted.Hash())
-	isHeaderInStorage := err == nil
-	if isHeaderInStorage {
-		log.Debug("intercepted block header already processed")
+	isHeaderOkForProcessing := hi.hdrInterceptorBase.headerValidator.CheckHeaderHandlerValid(hdrIntercepted.Header)
```
Comment (sasurobert) [inline process/block/interceptors/headerInterceptor.go]: hdrInterceptorBase could be made as only a reference to the base that have a simpler hi.headerValidator.CheckHeaderValid(...). Like *baseProcessor is in shardProcessor.

File: process/block/preprocess/transactions.go [hunk 6]
```diff
@@ -407,7 +402,7 @@ func (txs *transactions) CreateAndProcessMiniBlock(sndShardId, dstShardId uint32
 		return nil, process.ErrTimeIsOut
 	}
 
-	log.Debug(fmt.Sprintf("time elapsed to ordered %d txs: %v sec\n", len(orderedTxes), timeAfter.Sub(timeBefore).Seconds()))
+	log.Info(fmt.Sprintf("time elapsed to ordered %d txs: %v sec\n", len(orderedTxes), timeAfter.Sub(timeBefore).Seconds()))
```
Comment (sasurobert) [inline process/block/preprocess/transactions.go]: we should use this as debug.

File: process/block/preprocess/transactions.go [hunk 7]
```diff
@@ -434,7 +429,7 @@ func (txs *transactions) CreateAndProcessMiniBlock(sndShardId, dstShardId uint32
 		)
 
 		if err != nil {
-			log.Error(err.Error())
+			log.Info(err.Error())
```
Comment (sasurobert) [inline process/block/preprocess/transactions.go]: leave it to log.error

File: process/block/preprocess/transactions.go [hunk 8]
```diff
@@ -451,6 +446,10 @@ func (txs *transactions) CreateAndProcessMiniBlock(sndShardId, dstShardId uint32
 		}
 	}
 
+	if len(miniBlock.TxHashes) < 3 && len(miniBlock.TxHashes) > 0 {
```
Comment (sasurobert) [inline process/block/preprocess/transactions.go]: delete bubu

File: process/dataValidators/nilHeaderHandlerProcessValidator.go [hunk 9]
```diff
@@ -0,0 +1,17 @@
+package dataValidators
+
+import "github.com/ElrondNetwork/elrond-go/data"
+
+// nilHeaderHandlerProcessValidator represents a header handler validator that doesn't check the validity of provided headerHandler
+type nilHeaderHandlerProcessValidator struct {
```
Comment (sasurobert) [inline process/dataValidators/nilHeaderHandlerProcessValidator.go]: too long name: suggestion HeaderValidator

File: process/dataValidators/nilHeaderHandlerProcessValidator.go [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+}
+
+// NewNilHeaderHandlerProcessValidator creates a new nil header handler validator instance
+func NewNilHeaderHandlerProcessValidator() (*nilHeaderHandlerProcessValidator, error) {
+	return &nilHeaderHandlerProcessValidator{}, nil
+}
+
+// CheckHeaderHandlerValid is a nil implementation that will return true
+func (nhhv *nilHeaderHandlerProcessValidator) CheckHeaderHandlerValid(headerHandler data.HeaderHandler) bool {
```
Comment (sasurobert) [inline process/dataValidators/nilHeaderHandlerProcessValidator.go]: suggestion to rename: IsHeaderValid

File: process/dataValidators/nilTxHandlerProcessValidator.go [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+type nilTxHandlerProcessValidator struct {
+}
+
+// NewNilTxHandlerProcessValidator creates a new nil tx handler validator instance
+func NewNilTxHandlerProcessValidator() (*nilTxHandlerProcessValidator, error) {
+	return &nilTxHandlerProcessValidator{}, nil
+}
+
+// CheckTxHandlerValid is a nil implementation that will return true
+func (nthv *nilTxHandlerProcessValidator) CheckTxHandlerValid(txHandler data.TransactionHandler) bool {
```
Comment (sasurobert) [inline process/dataValidators/nilTxHandlerProcessValidator.go]: rename suggestion IsTxValid

File: process/dataValidators/nilTxHandlerProcessValidator.go [hunk 12]
```diff
@@ -0,0 +1,17 @@
+package dataValidators
+
+import "github.com/ElrondNetwork/elrond-go/data"
+
+// nilTxHandlerProcessValidator represents a tx handler validator that doesn't check the validity of provided txHandler
+type nilTxHandlerProcessValidator struct {
```
Comment (sasurobert) [inline process/dataValidators/nilTxHandlerProcessValidator.go]: name too long: nilTxValidator should be enough

File: process/interface.go [hunk 13]
```diff
@@ -31,6 +31,16 @@ type TxTypeHandler interface {
 	ComputeTransactionType(tx data.TransactionHandler) (TransactionType, error)
 }
 
+// TxHandlerProcessValidator can determine if a provided transaction handler is valid or not from the process point of view
+type TxHandlerProcessValidator interface {
+	CheckTxHandlerValid(txHandler data.TransactionHandler) bool
+}
+
+// HeaderHandlerProcessValidator can determine if a provided header handler is valid or not from the process point of view
+type HeaderHandlerProcessValidator interface {
```
Comment (sasurobert) [inline process/interface.go]: interface name too long: HeaderValidator should be enough. Same for transactions as well: TransactionValidator. Should rename function to IsTransactionValid and IsHeaderValid

File: process/sync/basicForkDetector.go [hunk 14]
```diff
@@ -113,12 +113,6 @@ func (bfd *basicForkDetector) checkBlockValidity(header data.HeaderHandler, stat
 	if int64(header.GetRound()) > bfd.rounder.Index() {
 		return ErrHigherRoundInBlock
 	}
-	if header.GetNonce() == bfd.fork.checkpointNonce {
```
Comment (sasurobert) [inline process/sync/basicForkDetector.go]: I think this should be remove by the other PR.

File: process/transaction/process_test.go [hunk 15]
```diff
@@ -397,8 +397,6 @@ func TestTxProcessor_GetSameAccountShouldWork(t *testing.T) {
 //------- checkTxValues
 
 func TestTxProcessor_CheckTxValuesHigherNonceShouldErr(t *testing.T) {
-	t.Skip()
-
 	adr1 := mock.NewAddressMock([]byte{65})
```
Comment (sasurobert) [inline process/transaction/process_test.go]: add t.Parallel, for the 2 other tests as well.

File: integrationTests/multiShard/block/executingMiniblocksSc_test.go (hunk 1 above)
Comment (iulianpascalau) [reply integrationTests/multiShard/block/executingMiniblocksSc_test.go]: Yes, @BeniaminDrasovean is doing this. I just patch the file in order to get the task done.

File: integrationTests/multiShard/block/executingMiniblocks_test.go (hunk 3 above)
Comment (iulianpascalau) [reply integrationTests/multiShard/block/executingMiniblocks_test.go]: Agree, will be completely removed from this file when the refactoring is done. This func has code from TestNodeProcessor.SyncBlock()

File: integrationTests/multiShard/block/executingMiniblocks_test.go (hunk 4 above)
Comment (iulianpascalau) [reply integrationTests/multiShard/block/executingMiniblocks_test.go]: Yes, but in the mess of this test file, n.miniblocks returned the miniblocks in the same order they were retrieved, leading to sync errors. getBlockBody actually get the miniblocks in the same order as described in the header. After refactoring is done, getBlockBody will be discarded, having the much more better func TestProcessorNode.SyncBlock() that does the same thing in a more elegant manner.

File: process/block/interceptors/headerInterceptor.go (hunk 5 above)
Comment (iulianpascalau) [reply process/block/interceptors/headerInterceptor.go]: hdrInterceptorBase is scheduled to be removed from the project.

File: process/block/preprocess/transactions.go (hunk 6 above)
Comment (iulianpascalau) [reply process/block/preprocess/transactions.go]: :+1:

File: process/block/preprocess/transactions.go (hunk 7 above)
Comment (iulianpascalau) [reply process/block/preprocess/transactions.go]: ok

File: process/block/preprocess/transactions.go (hunk 8 above)
Comment (iulianpascalau) [reply process/block/preprocess/transactions.go]: I can not believe this got through! Surely will be deleted!

File: process/dataValidators/nilHeaderHandlerProcessValidator.go (hunk 9 above)
Comment (iulianpascalau) [reply process/dataValidators/nilHeaderHandlerProcessValidator.go]: Would keep process for now as the interceptors will be refactored as a whole. Maybe will need a HeaderHandlerDataValidator. About handler part, yes, it can be removed.

File: process/dataValidators/nilHeaderHandlerProcessValidator.go (hunk 10 above)
Comment (iulianpascalau) [reply process/dataValidators/nilHeaderHandlerProcessValidator.go]: Great! Will change!

File: process/dataValidators/nilTxHandlerProcessValidator.go (hunk 11 above)
Comment (iulianpascalau) [reply process/dataValidators/nilTxHandlerProcessValidator.go]: Will change

File: process/block/preprocess/transactions.go (hunk 8 above)
Comment (iulianpascalau) [reply process/block/preprocess/transactions.go]: Deleted by @SebastianMarian

File: integrationTests/multiShard/block/executingMiniblocksSc_test.go (hunk 2 above)
Comment (iulianpascalau) [reply integrationTests/multiShard/block/executingMiniblocksSc_test.go]: Good catch

Comment (iulianpascalau) [reply integrationTests/multiShard/block/executingMiniblocksSc_test.go]: Done

File: process/block/preprocess/transactions.go (hunk 6 above)
Comment (iulianpascalau) [reply process/block/preprocess/transactions.go]: Done

File: process/block/preprocess/transactions.go (hunk 7 above)
Comment (iulianpascalau) [reply process/block/preprocess/transactions.go]: Done

File: process/dataValidators/nilHeaderHandlerProcessValidator.go (hunk 9 above)
Comment (iulianpascalau) [reply process/dataValidators/nilHeaderHandlerProcessValidator.go]: Renmed to HeaderValidator

File: process/dataValidators/nilHeaderHandlerProcessValidator.go (hunk 10 above)
Comment (iulianpascalau) [reply process/dataValidators/nilHeaderHandlerProcessValidator.go]: Done

File: process/dataValidators/nilTxHandlerProcessValidator.go (hunk 11 above)
Comment (iulianpascalau) [reply process/dataValidators/nilTxHandlerProcessValidator.go]: Done

File: process/dataValidators/nilTxHandlerProcessValidator.go (hunk 12 above)
Comment (iulianpascalau) [reply process/dataValidators/nilTxHandlerProcessValidator.go]: Done

File: process/interface.go (hunk 13 above)
Comment (iulianpascalau) [reply process/interface.go]: HeaderValidator and TxValidator

File: process/transaction/process_test.go (hunk 15 above)
Comment (iulianpascalau) [reply process/transaction/process_test.go]: Added t.Parallel on the entire test file

File: integrationTests/multiShard/block/executingMiniblocks_test.go (hunk 3 above)
Comment (sasurobert) [reply integrationTests/multiShard/block/executingMiniblocks_test.go]: Still you should rename the function until the refactor is done. Just add a longer name, createBlockBodyFromOrderedMiniblocks or something similar.

Comment (iulianpascalau) [reply integrationTests/multiShard/block/executingMiniblocks_test.go]: ok

File: process/sync/basicForkDetector.go (hunk 14 above)
Comment (SebastianMarian) [reply process/sync/basicForkDetector.go]: Done

File: integrationTests/multiShard/block/executingMiniblocks_test.go (hunk 3 above)
Comment (SebastianMarian) [reply integrationTests/multiShard/block/executingMiniblocks_test.go]: Done

File: process/errors.go [hunk 16]
```diff
@@ -408,3 +405,9 @@ var ErrUnknownBlockType = errors.New("block type is unknown")
 
 // ErrMissingPreProcessor signals that required pre processor is missing
 var ErrMissingPreProcessor = errors.New("pre processor is missing")
+
+// ErrNilTxValidator signals that a nil tx handler validator has been provided
+var ErrNilTxValidator = errors.New("nil tx handler validator provided")
```
Comment (BeniaminDrasovean) [inline process/errors.go]: The name and comment of this error is confusing because of the nilTxValidator struct. Does it mean that the tx validator is nil, or the tx validator is nilTxValidator? Same for the next error.
